### ISSUE TITLE
Usage with `MonacoWebpackPlugin`? 

### ISSUE THREAD

**Issue by dotansimha** [pengx17/monaco-yaml#20]
I tried multiple solutions, mostly around trying to integrate the worker loading phase into an existing webpack, and it didn't work for me (ForgeinModule errors, or issues with `monaco` is undefined).

I managed to make it work with `MonacoWebpackPlugin` only when I modify the built-in languages array of `MonacoWebpackPlugin` before loading it:

```js
const { languagesArr } = require('monaco-editor-webpack-plugin/out/languages');

const yamlLang = languagesArr.find(t => t.label === 'yaml');

yamlLang.entry = [
  yamlLang.entry,
  '../../monaco-yaml/esm/monaco.contribution'
];
yamlLang.worker = {
  id: 'vs/language/yaml/yamlWorker',
  entry: '../../monaco-yaml/esm/yaml.worker.js'
}

const MonacoWebpackPlugin = require('monaco-editor-webpack-plugin');

// later in your code, register MonacoWebpackPlugin in your webpack plugins
```

This is the only way I managed to get the worker loaded correctly with `MonacoWebpackPlugin`, because all registration, imports of `monaco.contribution` , loading the worker, and configuring the `MonacoEnvironment ` is managed by the plugin automatically. 

Maybe this should be documented? or improved somehow?

**Comment by yazaabed** [comment]
Thanks for your workaround to make this working. I also have another way to do this, by loading the monaco-editor by worker-loader or worker-plugin in webpack and override the monaco environment object. Check the example, notice the `MonacoEnvironment`:

```
// Lazy load monaco-editor by webpack dynamic-imports.
const monaco = await import(/* webpackChunkName: "editor.main" */"monaco-editor/esm/vs/editor/editor.api");

// @note: This is lazy-loading for css files, will be compiled at
// angular-cli side. And no need for typescript to understand this.
// @ts-ignore
const styles = await import(/* webpackChunkName: "editor.styles" */"style-loader?injectType=lazyStyleTag!@monaco-editor-styles")

// Load styles on-demand.
this.styles = styles;
this.styles.use();

// Load monaco-yaml plugin
if (this.options.language === "yaml") {
    // @ts-ignore
    await import(/* webpackChunkName: "yaml.contribution" */"monaco-yaml/esm/monaco.contribution");

    // @ts-ignore
    const yamlWorkerUrl = await import(/* webpackChunkName: "yamlWorker.url" */"worker-plugin/loader?name=yaml&esModule!monaco-yaml/esm/yaml.worker");

    // monaco-editor add `MonacoEnvironment` to window object.
    const getWorkerUrl = (window as any).MonacoEnvironment.getWorkerUrl;

    // Add yaml worker to other workers provided by monaco-editor.
    (window as any).MonacoEnvironment.getWorkerUrl = (module: string, label: string) => {
        if (label === "yaml") {
            return yamlWorkerUrl.default;
        }

        return getWorkerUrl(module, label);
    };
}
```

**Comment by dotansimha** [comment]
Thanks @yazaabed , I tried this approach with the loaders, but I got tons of errors about loading the worker, and looks like my Webpack setup was missing something. I'm using Docusaurous v2 as base setup, so maybe the default settings of it might effect it as well.

**Comment by yazaabed** [comment]
> Thanks @yazaabed , I tried this approach with the loaders, but I got tons of errors about loading the worker, and looks like my Webpack setup was missing something. I'm using Docusaurous v2 as base setup, so maybe the default settings of it might effect it as well.

I can help with it if you are intrested in make it work with loader. Are you using webpack v2?

**Comment by dotansimha** [comment]
@yazaabed I think it might be better to consider to add an example for using it with `MonacoWebpackPlugin`? it seems like it's the simplest way to integrate it easily (especially if you can only extend the Webpack configuration and you are not writing it from scratch).

The project I applied it to is here: https://github.com/dotansimha/graphql-code-generator/blob/master/website/monaco-plugin.js , it's based on Docusaurous v2, with webpack 4).

**Comment by yazaabed** [comment]
> @yazaabed I think it might be better to consider to add an example for using it with `MonacoWebpackPlugin`? it seems like it's the simplest way to integrate it easily (especially if you can only extend the Webpack configuration and you are not writing it from scratch).
> 
> The project I applied it to is here: https://github.com/dotansimha/graphql-code-generator/blob/master/website/monaco-plugin.js , it's based on Docusaurous v2, with webpack 4).

Yes I agree your way to fix this is much simpler than my way to fix it. Maybe you can create a PR and add an example in the examples folder.

**Comment by yazaabed** [comment]
In the new version from web pack plugin, a newly option added called customLanguages which can be used to define the yaml editor. As we now support the new version of Monaco-editor.

**Comment by JounQin** [comment]
@yazaabed Do you have any reference or example about the `customLanguages` option? I have no idea what it is.

**Comment by NickTomlin** [comment]
I've gotten this to work with the following config: 

```javascript
module.exports = {
  webpack: {
    plugins: [
      new MonacoWebpackPlugin({
        languages: ['json', 'yaml', 'graphql'],
        customLanguages: [{
          label: 'yaml',
          // TODO: restore once https://github.com/pengx17/monaco-yaml/pull/48 is merged
          entry: '../../@undefined-moe/monaco-yaml/lib/esm/monaco.contribution',
          worker: {
            id: 'vs/language/yaml/yamlWorker',
            // TODO: restore once https://github.com/pengx17/monaco-yaml/pull/48 is merged
            entry: '../../@undefined-moe/monaco-yaml/lib/esm/yaml.worker.js'
          }
        }]
      })
    ]
  }
}
```

⚠️  There's currently an issue with more recent versions of monaco that is fixed by #48. My example uses a forked module that will not be up to date. Replace with the non namespaced version of `monaco-yaml` once that is merged ⚠️

**Comment by JounQin** [comment]
There is a `globalAPI: true` available in `monaco-editor-webpack-plugin@3.1.0` now, this could be a workaround before #48.
